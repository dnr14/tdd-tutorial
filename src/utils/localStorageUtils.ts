export const _TODO_KEY = 'todo_list';
