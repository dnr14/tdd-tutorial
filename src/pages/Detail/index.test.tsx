import { screen, render, fireEvent } from '@testing-library/react';
import 'jest-styled-components';
import { Detail } from '.';
import { _TODO_KEY } from 'utils/localStorageUtils';
import { createMemoryHistory } from 'history';
import { Route, Router, useLocation } from 'react-router-dom';

const _setLocalStorage = () => {
  localStorage.setItem(_TODO_KEY, JSON.stringify(['todo1', 'todo2']));
};

const TestCompo = () => {
  const { pathname } = useLocation();
  return <div>{pathname}</div>;
};

const HIS = ({ path = '/detail/1' }: { path?: string }) => {
  const history = createMemoryHistory();
  history.push(`${path}`);

  return (
    <Router history={history}>
      <TestCompo />
      <Route path={'/detail/:id'}>
        <Detail />
      </Route>
    </Router>
  );
};

beforeAll(() => {
  _setLocalStorage();
});

afterAll(() => {
  localStorage.clear();
});

describe('<Detail>', () => {
  it('<Detail> 컴포넌트 랜더링 확인', () => {
    const { container } = render(<HIS />);

    const todoItem2 = screen.getByText('todo2');
    expect(todoItem2).toBeInTheDocument();
    const button = screen.getByText('삭제');
    expect(button).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('redirect to Not Found page if todo id is wrong', () => {
    render(<HIS path="/detail/99" />);
    const url = screen.getByText('/404');
    expect(url).toBeInTheDocument();
  });

  it('delete a ToDo and redirect to the List page', () => {
    render(<HIS />);

    const url = screen.getByText('/detail/1');
    expect(url).toBeInTheDocument();

    const button = screen.getByText('삭제');
    fireEvent.click(button);

    expect(JSON.parse(localStorage.getItem(_TODO_KEY) as string)).not.toContain('todo2');
    expect(url.textContent).toBe('/');
  });
});
