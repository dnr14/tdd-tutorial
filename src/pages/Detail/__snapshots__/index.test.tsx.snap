// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Detail> <Detail> 컴포넌트 랜더링 확인 1`] = `
.c2 {
  text-align: center;
  padding: 10px 20px;
  border-radius: 8px;
  cursor: pointer;
  background-color: #FF1744;
}

.c2:hover {
  background-color: #F01440;
}

.c2:active {
  box-shadow: 5px 5px 10px rgba(0,0,0,0.2);
}

.c3 {
  color: #fff;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #fff;
  padding: 20px;
  border-radius: 8px;
  box-shadow: 5px 5px 10px rgba(0,0,0,0.2);
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  min-width: 350px;
  height: 350px;
  overflow-y: auto;
  border: 1px solid #bdbdbd;
  margin-bottom: 20px;
  padding: 10px;
}

<div>
  <div>
    /detail/1
  </div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      todo2
    </div>
    <div
      class="c2"
      data-testid="button-container"
    >
      <div
        class="c3"
      >
        삭제
      </div>
    </div>
  </div>
</div>
`;
